cmake/extensions.cmake

```diff
@@ -371,6 +371,93 @@ function(generate_inc_file_for_target
   generate_inc_file_for_gen_target(${target} ${source_file} ${generated_file} ${generated_target_name} ${ARGN})
 endfunction()
 
+function(target_sources_codegen target)
+  foreach(arg ${ARGN})
+    # Generated file must be generated to the current binary directory.
+    # Otherwise this would trigger CMake issue #14633:
+    # https://gitlab.kitware.com/cmake/cmake/issues/14633
+    if(IS_ABSOLUTE ${arg})
+      set(template_file ${arg})
+      get_filename_component(generated_file_name ${arg} NAME)
+      set(generated_file ${CMAKE_CURRENT_BINARY_DIR}/${generated_file_name})
+    else()
+      set(template_file ${CMAKE_CURRENT_SOURCE_DIR}/${arg})
+      set(generated_file ${CMAKE_CURRENT_BINARY_DIR}/${arg})
+    endif()
+
+    if(IS_DIRECTORY ${template_file})
+      message(FATAL_ERROR "target_sources_codegen() was called on a directory")
+    endif()
+
+    # Generate file from template
+    message(STATUS "Will generate ${generated_file}")
+    message(STATUS "         from ${template_file}")
+    add_custom_command(
+      OUTPUT ${generated_file}
+      DEPENDS ${template_file}
```

Comment:
We need to support users specifying dependencies.

E.g. the generation of offsets.h depends on the target 'offsets'.

One way to support this would be to support an optional 'DEPENDS' argument, e.g. through through the use of cmake_parse_arguments. Will see if I have time to prototype this after porting gen_offset_header.py.

Reply:
cmake_parse_arguments works. What about:

  zephyr_sources_codegen(file.c ARGS myarg=${xxx} yourarg==123 DEPENDS my_file.h)

Reply:
Yep, that would work.